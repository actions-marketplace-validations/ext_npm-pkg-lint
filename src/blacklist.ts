import blacklist from "./blacklist/index";

export function setupBlacklist(name: string): void {
	/* eslint-disable-next-line security/detect-non-literal-regexp -- expected to be non-literal */
	blacklist.push(new RegExp(`^${name}-\\d+\\.\\d+\\.\\d+\\.tgz$`));
}

export function isBlacklisted(filename: string): boolean {
	return blacklist.some((entry) => filename.match(entry));
}
