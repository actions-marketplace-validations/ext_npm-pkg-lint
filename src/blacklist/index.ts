import emacs from "./emacs";
import eslint from "./eslint";
import fimbullinter from "./fimbullinter";
import git from "./git";
import jest from "./jest";
import node from "./node";
import prettier from "./prettier";
import typescript from "./typescript";
import windows from "./windows";
import { directory, extension, filename, rcfile } from "./helpers";

const blacklist: RegExp[] = [
	/* tests in general */
	directory(".nyc_output"),
	directory("coverage"),
	extension(".lcov"),
	extension("[.-]junit.xml"),
	filename(".istanbul.yml"),
	filename("junit.xml"),
	filename("lcov.info"),
	filename(".airtap.yml"),
	rcfile(".nycrc"),

	/* tests */
	directory("examples?"),
	directory("perf"),
	directory("tests?"),
	filename("cypress.json"),
	filename("karma.config.js"),
	filename("mocha.opts"),
	filename("protractor.config.js"),
	filename("test.js"),
	filename("testem.json"),
	filename("testem.yml"),
	filename("wallaby.config.js"),
	filename("wallaby.js"),

	/* linting */
	filename(".htmlvalidate.json"),
	filename(".jshintignore"),
	filename(`tslint.json`),
	filename(`tslint.yaml`),
	filename(".jscs.json"),
	rcfile(`.jscsrc`),
	rcfile(`.jshintrc`),

	/* toolchains */
	filename("rollup.config.js"),
	filename("webpack.config.js"),

	/* task runners */
	filename("[gG]runtfile.js"),
	filename("[gG]ulpfile.[jt]s"),
	filename("dangerfile.[jt]s"),
	filename("gulpfile.babel.js"),
	filename("gulpfile.esm.js"),

	/* ci */
	/^\.azure-.*$/,
	directory(".circleci"),
	directory(".github"),
	directory(".gitlab"),
	directory(".?zuul.d"),
	filename(".coveralls.yml"),
	filename(".gitlab-ci.yml"),
	filename(".travis.yml"),
	filename(".?zuul.ya?ml"),
	filename("ISSUE_TEMPLATE.md"),
	filename("PULL_REQUEST_TEMPLATE.md"),
	filename("appveyor.yml"),

	/* editor specific files */
	/(^|\/)\.sw[a-p]/ /* vim */,
	directory(".vscode"),
	extension(".sublime-project"),
	extension(".sublime-workspace"),
	filename(".vscodeignore"),
	filename("Session.vim"),
	filename("Sessionx.vim"),
	directory(".idea"),
	filename(".tm_properties") /* textmate */,

	/* OS */
	directory(".AppleDouble"),
	directory(".DS_Store"),
	directory(".LSOverride"),

	/* logs */
	directory("logs"),
	extension(".log"),

	/* runtime */
	directory("pids"),
	extension(".pid"),
	extension(".pid.lock"),
	extension(".seed"),

	/* misc */
	/^[Jj]akefile(\.js)?$/,
	extension(".bak"),
	extension(".swp"),
	extension(".tgz"),
	extension(".tmp"),
	filename(".env"),
	filename(".svgo.yml"),
	filename(".editorconfig"),
	filename(".ext-prepush"),
	filename(".tonic_example.js"),
	filename(".flowconfig"),
	filename(".cardinalrc"),
	filename("bower.json"),
	filename("commitlint.config.js") /* commitlint */,
	filename("composer.json"),
	filename("netlify.toml"),
	filename("release.config.js") /* semantic release */,

	...emacs,
	...eslint,
	...fimbullinter,
	...git,
	...jest,
	...node,
	...prettier,
	...typescript,
	...windows,
];

export default blacklist;
