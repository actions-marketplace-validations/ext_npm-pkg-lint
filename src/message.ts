export interface Message {
	ruleId: string;
	severity: number;
	message: string;
	line: number;
	column: number;
}
