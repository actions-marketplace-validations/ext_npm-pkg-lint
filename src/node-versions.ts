interface NodeVersionDescriptor {
	/** If set this version is EOL since given date */
	eol?: string;
}

export const nodeVersions: Array<[string, NodeVersionDescriptor]> = [
	["0.10.x", { eol: "2016-10-31" }],
	["0.12.x", { eol: "2016-12-31" }],
	["4.x.x", { eol: "2018-04-30" }],
	["5.x.x", { eol: "2016-06-30" }],
	["6.x.x", { eol: "2019-04-30" }],
	["7.x.x", { eol: "2017-06-30" }],
	["8.x.x", { eol: "2019-12-31" }],
	["9.x.x", { eol: "2018-06-30" }],
	["10.x.x", { eol: "2021-04-30" }],
	["11.x.x", { eol: "2019-06-01" }],
	["12.x.x", { eol: "2022-04-30" }],
	["13.x.x", { eol: "2020-06-01" }],
	["14.x.x", { eol: "2023-04-30" }],
	["15.x.x", { eol: "2021-06-01" }],
];
