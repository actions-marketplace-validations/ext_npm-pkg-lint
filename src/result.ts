import { Message } from "./message";

export interface Result {
	messages: Message[];
	filePath: string;
	errorCount: number;
	warningCount: number;
}
