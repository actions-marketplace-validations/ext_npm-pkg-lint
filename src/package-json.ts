import PackageJson from "./types/package-json";
import { Message } from "./message";
import { Result } from "./result";
import { nonempty, present, typeArray, typeString, validUrl } from "./validators";
import { isDisallowedDependency } from "./rules/disallowed-dependency";

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
type validator = (key: string, value: any) => void;

const fields: Record<string, validator[]> = {
	description: [present, typeString, nonempty],
	keywords: [present, typeArray, nonempty],
	homepage: [present, typeString, validUrl],
	bugs: [present, validUrl],
	license: [present, typeString, nonempty],
	author: [present, nonempty],
	repository: [present, validUrl],
};

function verifyFields(pkg: PackageJson): Message[] {
	const messages: Message[] = [];

	for (const [field, validators] of Object.entries(fields)) {
		try {
			for (const validator of validators) {
				validator(field, pkg[field]);
			}
		} catch (error) {
			messages.push({
				ruleId: "package-json-fields",
				severity: 2,
				message: error.message,
				line: 1,
				column: 1,
			});
		}
	}

	return messages;
}

function verifyDependencies(pkg: PackageJson): Message[] {
	const messages: Message[] = [];

	for (const dependency of Object.keys(pkg.dependencies || {})) {
		if (isDisallowedDependency(dependency)) {
			messages.push({
				ruleId: "disallowed-dependency",
				severity: 2,
				message: `${dependency} should be a devDependency`,
				line: 1,
				column: 1,
			});
		}
	}

	return messages;
}

export async function verifyPackageJson(pkg: PackageJson, filePath: string): Promise<Result[]> {
	const messages: Message[] = [...verifyFields(pkg), ...verifyDependencies(pkg)];

	if (messages.length === 0) {
		return [];
	}

	return [
		{
			messages,
			filePath,
			errorCount: messages.length,
			warningCount: 0,
			fixableErrorCount: 0,
			fixableWarningCount: 0,
		},
	];
}
