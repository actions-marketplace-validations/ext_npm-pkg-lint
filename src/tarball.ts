import tar, { FileStat } from "tar";
import PackageJson from "./types/package-json";
import { isBlacklisted } from "./blacklist";
import { Message } from "./message";
import { Result } from "./result";

interface RequiredFile {
	field: string;
	ruleId: string;
	filename: string;
}

export async function getFileList(filename: string): Promise<string[]> {
	const entries: FileStat[] = [];
	await tar.list({
		file: filename,
		strict: true,
		onentry: (entry: FileStat) => entries.push(entry),
	});
	return entries.map((entry) => {
		const filename = (entry.path as unknown) as string;
		return filename.replace(/^package\//, "");
	});
}

export async function blacklistedFiles(filelist: string[]): Promise<string[]> {
	return filelist.filter(isBlacklisted);
}

function normalizeRequiredFiles(src: string | string[] | Record<string, string>): string[] {
	if (typeof src === "string") {
		return [src];
	} else if (Array.isArray(src)) {
		return src;
	} else {
		return Object.values(src);
	}
}

function* yieldRequiredFiles(
	src: string | string[] | Record<string, string>,
	template: Pick<RequiredFile, "field" | "ruleId">
): Generator<RequiredFile> {
	const files = normalizeRequiredFiles(src);
	for (const filename of files) {
		yield { ...template, filename };
	}
}

function* requiredFiles(pkg: PackageJson): Generator<RequiredFile> {
	if (pkg.main) {
		yield* yieldRequiredFiles(pkg.main, { field: "main", ruleId: "no-missing-main" });
	}
	if (pkg.browser) {
		yield* yieldRequiredFiles(pkg.browser, { field: "browser", ruleId: "no-missing-main" });
	}
	if (pkg.module) {
		yield* yieldRequiredFiles(pkg.module, { field: "module", ruleId: "no-missing-main" });
	}
	if (pkg["jsnext:main"]) {
		yield* yieldRequiredFiles(pkg["jsnext:main"], {
			field: "jsnext:main",
			ruleId: "no-missing-main",
		});
	}
	if (pkg.typings) {
		yield* yieldRequiredFiles(pkg.typings, { field: "typings", ruleId: "no-missing-typings" });
	}
	if (pkg.bin) {
		yield* yieldRequiredFiles(pkg.bin, { field: "bin", ruleId: "no-missing-binary" });
	}
	if (pkg.man) {
		yield* yieldRequiredFiles(pkg.man, { field: "man", ruleId: "no-missing-man" });
	}
}

export async function verifyTarball(pkg: PackageJson, filePath: string): Promise<Result> {
	const messages: Message[] = [];
	const filelist = await getFileList(filePath);

	for (const filename of await blacklistedFiles(filelist)) {
		messages.push({
			ruleId: "no-disallowed-files",
			severity: 2,
			message: `${filename} is not allowed in tarball`,
			line: 1,
			column: 1,
		});
	}

	for (const requiredFile of requiredFiles(pkg)) {
		if (!filelist.includes(requiredFile.filename)) {
			messages.push({
				ruleId: requiredFile.ruleId,
				severity: 2,
				message: `${requiredFile.filename} (pkg.${requiredFile.field}) is not present in tarball`,
				line: 1,
				column: 1,
			});
		}
	}

	return {
		messages,
		filePath,
		errorCount: messages.length,
		warningCount: 0,
	};
}
