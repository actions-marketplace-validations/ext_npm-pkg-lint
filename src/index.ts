/* eslint-disable no-console, no-process-exit -- this is a cli tool */

import { existsSync, promises as fs } from "fs";
import path from "path";
import { ArgumentParser } from "argparse";
import findUp from "find-up";
import stylish from "@html-validate/stylish";
import { setupBlacklist } from "./blacklist";
import { verify } from "./verify";
import PackageJson from "./types/package-json";
import { tarballLocation } from "./tarball-location";
import { getFileContent } from "./tarball";

/* eslint-disable-next-line @typescript-eslint/no-var-requires */
const { version } = require("../package.json");

interface ParsedArgs {
	pkgfile: string;
	tarball?: string;
}

interface GetPackageJsonResults {
	pkg?: PackageJson;
	pkgPath?: string;
}

async function getPackageJson(args: ParsedArgs): Promise<GetPackageJsonResults> {
	/* get from explicit path passed as argument */
	if (args.pkgfile) {
		return {
			pkg: JSON.parse(await fs.readFile(args.pkgfile, "utf-8")),
			pkgPath: args.pkgfile,
		};
	}

	/* extract package.json from explicit tarball location */
	if (args.tarball) {
		const contents = await getFileContent(args.tarball, ["package.json"]);
		return {
			pkg: JSON.parse(contents["package.json"].toString("utf-8")),
			pkgPath: path.join(args.tarball, "package.json"),
		};
	}

	/* try to locate package.json from file structure */
	const pkgPath = await findUp("package.json");
	if (pkgPath) {
		const relPath = path.relative(process.cwd(), pkgPath);
		return {
			pkg: JSON.parse(await fs.readFile(relPath, "utf-8")),
			pkgPath: relPath,
		};
	}

	return {};
}

async function run(): Promise<void> {
	const parser = new ArgumentParser({
		description: "npm package linter",
	});

	parser.add_argument("-v", "--version", { action: "version", version });
	parser.add_argument("-t", "--tarball", { help: "specify tarball location" });
	parser.add_argument("-p", "--pkgfile", { help: "specify package.json location" });

	const args: ParsedArgs = parser.parse_args();
	const { pkg, pkgPath } = await getPackageJson(args);

	if (!pkg) {
		console.error("Failed to locate package.json and no location was specificed with `--pkgfile'");
		process.exit(1);
	}

	const tarball = args.tarball ?? tarballLocation(pkg, pkgPath);
	if (!existsSync(tarball)) {
		console.error(`"${tarball}" does not exist, did you forget to run \`npm pack'?`);
		process.exit(1);
	}

	setupBlacklist(pkg.name);

	const results = await verify(pkg, pkgPath, tarball);
	const output = stylish(results);
	process.stdout.write(output);

	const totalErrors = results.reduce((sum, result) => {
		return sum + result.errorCount;
	}, 0);

	process.exit(totalErrors > 0 ? 1 : 0);
}

run();
