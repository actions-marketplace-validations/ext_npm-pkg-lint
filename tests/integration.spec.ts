jest.unmock("tar");

import { promises as fs } from "fs";
import path from "path";
import glob from "glob";
import execa from "execa";
import { verifyTarball } from "../src/tarball";
import PackageJson from "../src/types/package-json";

const ROOT_DIRECTORY = path.resolve(path.join(__dirname, ".."));
const FIXTURE_DIRECTORY = path.join(__dirname, "fixtures");

const fixtures = glob.sync(path.join(FIXTURE_DIRECTORY, "*")).map((fixture: string) => {
	return path.relative(FIXTURE_DIRECTORY, fixture);
});

async function npmPack(pkg: PackageJson, fixture: string): Promise<string> {
	const dir = path.join(FIXTURE_DIRECTORY, fixture);
	const tarball = path.join(dir, `${pkg.name}-${pkg.version}.tgz`);
	await execa("npm", ["pack"], { cwd: dir });
	return path.relative(ROOT_DIRECTORY, tarball);
}

it.each(fixtures)("%s", async (fixture) => {
	expect.assertions(1);
	const dir = path.join(FIXTURE_DIRECTORY, fixture);
	const pkg: PackageJson = JSON.parse(await fs.readFile(path.join(dir, "package.json"), "utf-8"));
	const tarball = await npmPack(pkg, fixture);
	const result = await verifyTarball(pkg, tarball);
	expect(result).toMatchSnapshot();
});
