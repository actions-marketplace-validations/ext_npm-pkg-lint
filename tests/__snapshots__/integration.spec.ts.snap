// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disallowed-files 1`] = `
Object {
  "errorCount": 1,
  "filePath": "tests/fixtures/disallowed-files/disallowed-files-1.0.0.tgz",
  "messages": Array [
    Object {
      "column": 1,
      "line": 1,
      "message": "my-lib.spec.js is not allowed in tarball",
      "ruleId": "no-disallowed-files",
      "severity": 2,
    },
  ],
  "warningCount": 0,
}
`;

exports[`missing-binary 1`] = `
Object {
  "errorCount": 1,
  "filePath": "tests/fixtures/missing-binary/missing-binary-1.0.0.tgz",
  "messages": Array [
    Object {
      "column": 1,
      "line": 1,
      "message": "bin.js (pkg.bin) is not present in tarball",
      "ruleId": "no-missing-binary",
      "severity": 2,
    },
  ],
  "warningCount": 0,
}
`;

exports[`missing-main 1`] = `
Object {
  "errorCount": 1,
  "filePath": "tests/fixtures/missing-main/missing-main-1.0.0.tgz",
  "messages": Array [
    Object {
      "column": 1,
      "line": 1,
      "message": "index.js (pkg.main) is not present in tarball",
      "ruleId": "no-missing-main",
      "severity": 2,
    },
  ],
  "warningCount": 0,
}
`;
